Add support for selecting tiles for multiple independent views

kring opened the pull request:
This is a big change, so I'm going to try to break it down here to hopefully make review more approachable.

## `TilesetViewGroup`

Represents a set of views that select tiles together. The actual views (`ViewState` instances) can change from frame to frame, and even the number of views in the group need not be consistent from one frame to the next.

However, the view group is "stateful" with respect to the selection algorithm. Which tiles were selected last frame in a view group affects which tiles will be selected this frame. This allows the algorithm to ensure that tiles don't blink out of existence from one frame to the next, and that unnecessary holes (missing tiles) are minimized.

The tiles selected for a view group are completely independent from other view groups. Two different view groups might cover a particular area of the model with different levels-of-detail, for example. When rendering a particular viewport, it's essential that only tiles from a single view group are rendered. View groups are useful when you have multiple viewports.

They might also be useful in some weirder circumstances. Rendering shadows using tiles selected by a view group with the perspective of a light source, perhaps? Or a view group at the destination of a camera flight to force those tiles to load before the camera arrives?

`TilesetViewGroup` can be default-constructed, and is owned by the user. Users are free to manage its lifetime however they see fit. It can be both moved and copied.

## `updateView` ➡️ `updateViewGroup`

Previously, users would call the `updateView` method on a `Tileset` and pass a list of frustums to use to select tiles. This will still work after this PR, and it will update the "default" view group. The default view group is owned by `Tileset` and can be obtained by calling `getDefaultViewGroup()`.

However, when using view groups, it is recommended to instead use the new `updateViewGroup` function. In addition to allowing a non-default view group to be updated, this function also does _not_ load tiles. That needs a little explanation.

View groups are not required to all be updated at the same rate. It's perfectly valid to update one view group at 60 FPS and another only rarely. Each time a view group is updated, it determines which tiles it would ideally like to load. The `Tileset` then uses a weighted round robin algorithm to give all view groups a fair chance to load tiles. The total number of tiles that may be in flight at any given time is, as before, limited by the `maximumSimultaneousTileLoads` property in `TilesetOptions`.

If we initiated tile loads after every `updateViewGroup`, then the first view group updated would likely use up all the tile load slots, because no other view groups would have any tiles to load yet. That's why `loadTiles` is a separate function on `Tileset` now, and must be called explicitly by the user when utilizing `updateViewGroup`. For backward compatibility, `updateView` calls `loadTiles` automatically after updating the default view group.

## `LoadedLinkedList` ➡️ `UnusedLinkedList`

Previously, Tiles that were traversed during `updateView` were added to an intrusive `LoadedLinkedList` on the `Tileset` called `_loadedTiles`. Each time a tile was visited, it was moved to the tail of the linked list. Since traversal always starts at the root, the root tile in the linked list marked the start of the tiles that were used in the most recent frame. Any tiles before this were not traversed last frame and so could be unloaded, with the least recently used closest to the head.

This is unworkable with multiple views, because we traverse the tile hierarchy multiple times. So the `LoadedLinkedList` is gone in this PR.

Instead, `TilesetViewGroup` now holds an `IntrusivePointer<const Tile>` for every tile that it visited during the last `updateViewGroup`. This intrusive pointer increments/decrements the `_referenceCount` field on the `Tile`. This is separate from and in addition to `_doNotUnloadSubtreeCount` that was added in #1107. It differs in two ways:

1. A non-zero `_referenceCount` keeps tile content from unloading. `_doNotUnloadSubtreeCount` only keeps subtrees from unloading and `Tile` instances from being deleted, but tile content is allowed to unload.
2. `_referenceCount` is not incremented on parent tiles when it is incremented on child tiles. The depth-first traversal will ensure that all ancestors of referenced children are also referenced.

When the `_referenceCount` goes to zero, the `Tile` is added to the tail of the new, intrusive `UnusedLinkedList`. If the `_referenceCount` goes back above zero, it is removed from the `UnusedLinkedList`. All tiles in this list can have their content unloaded, and the ones closer to the head are the least recently used.

This new mechanism for tracking unloadable tiles works well across view groups.

## `LoadedTileEnumerator`

The `LoadedLinkedList` was previously used to determine which tiles were unloadable and in what order they should be unloaded, as described above, but it had another purpose as well. It also allowed the complete set of "loaded" tiles to be enumerated. This was used for a few different things, notably adding or removing raster overlays to/from existing tiles.

So with the `LoadedLinkedList` now gone, we need a new way to enumerate the set of loaded tiles. In this PR, this job is handled by `LoadedTileEnumerator`. It works by traversing the tile hierarchy itself. Rather than traversing the entire tile hierarchy, though - which would be very slow - it only traverses tiles that either have loaded content or that have a `_doNotUnloadSubtreeCount` greater than zero (indicating that a descendant tile has loaded content).

This is probably a bit less efficient than the old linked list traversal, but it's still plenty fast for the types of use-cases where we use it.

## `TileLoadRequester`

I mentioned above that `Tileset` uses a weight round-robin algorithm to give multiple view groups a fair chance to load tiles. This happens via the new `TileLoadRequester` abstract base class.

Anything that wants to influence which tiles are loaded should inherit from `TileLoadRequester`. Currently that means two types: `TilesetViewGroup` and `TilesetHeightRequest` (which is created by a call to `sampleHeightMostDetailed`).

A `TileLoadRequester` may be registered with the `Tileset` by calling `Tileset::registerLoadRequester`. The requester offers two queues of tiles it wants to load, one that needs worker-thread loading and the other that needs main-thread loading. The two queues are exposed using a simple "has more?" / "get next!" interface, so the underlying mechanism for the storing the queue is up to the implementation. The requester also has `getWeight` method that determines the load priority of this requester versus the others. Here's the explanation of how the weight works, copied from the doxygen on the `getWeight` method:

> Most requesters should return a weight of 1.0. When all requesters have the same weight, they will all have an equal opportunity to load tiles. If one requester's weight is 2.0 and the rest are 1.0, that requester will have twice as many opportunities to load tiles as the others.
> 
> A very high weight will prevent all other requesters from loading tiles until this requester has none left to load. A very low weight (but above 0.0!) will allow all other requesters to finish loading tiles before this one starts.

## `TreeTraversalState`

One of the challenges in moving to multiple view groups is where to put the per-group "selection state". That is, whether a given tile was rendered, refined, culled, or kicked in the last frame. In order to do its job, `updateView` needs to know what decision was made _last frame_ for each tile that it visits. Previously, when we had only one view group, this was stored on the `Tile` itself.

A simple approach, like storing a map of "Tile -> Selection State" on the `TilesetViewGroup`, or a map of "View Group -> Selection State" on the `Tile`, would work just fine. Unfortunately, it would also be slow. I saw about a 20% increase in the time spent in `updateView` with this approach!

Instead, this PR uses a new class called `TreeTraversalState` to track the per-tile states during one traversal and access it during the next one. See #1129 for details about how this works.

## `CreditReferencer` (and changes to `CreditSystem`)

Credit / attribution tracking gets more complicated with the move to multiple view groups. Previously, `CreditSystem` worked in a sort of "immediate mode". Each frame (`updateView`), the credits applicable for the current update would be added to the `CreditSystem`. For convenience, `CreditSystem` employed a bit of bookkeeping so that it could report a) the credits active in the current frame, and b) the credits active last frame that are not active anymore.

This whole idea is unfortunately broken when we move to multiple viewports. View groups are not required to tick together, so there's no useful notion of a "frame" for credit purposes.

So `CreditSystem` has now moved to a "retained mode" pattern, where credits are explicitly enabled and disabled. A reference count tracks the number of times a credit has been enabled, and it won't disappear entirely until it has been disabled an equal number of time. The `CreditSystem` can produce a "snapshot" on request, which identifies the credits that are currently shown as well as the credits that have been hidden since the last snapshot.

Unfortunately, this makes credit management in `updateView` tricky. Conceptually, it's not too hard: when a tile is shown, enable its credits; when a tile is hidden, disable its credits. But it's more difficult in practice, because the set of credits associated with a `Tile` can change over time, such as when it's loaded or a new raster overlay is added. Ensuring that credit enable/disable calls are perfectly paired is error prone, and when it goes wrong it's difficult to debug.

That's why this PR introduces a new type, `CreditReferencer`. `CreditReferencer` exposes the old "immediate mode" pattern to the `TilesetViewGroup` while actually manipulating the retained reference counts that are now owned by `CreditSystem`. 

`TilesetViewGroup` has two `CreditReferencer` instances: `_previousFrameCredits` and `_currentFrameCredits`. 
In `TilesetViewGroup::finishFrame`, which happens at the end of `updateViewGroup`, all of the active credits added to the `_currentFrameCredits`, which indirectly enables them on the `CreditSystem`. Then, all of the credits in `_previousFrameCredits` are efficiently disabled. Finally, the two `CreditReferencer` instances are swapped.

Fixes #928 

To do:

* [x] Every call to `updateViewGroup` shouldn't return the same `ViewUpdateResult` from a field on the `Tileset`. Instead, the `ViewUpdateResult` could be stored in the `TilesetViewGroup`. Or perhaps these don't even need to be separate classes?
* [x] Check performance of the `std::unordered_map` based tile selection state tracking. My feeling is that this is going to be too slow.
* [x] Fix and reactivate some code that was temporarily disabled, such as `DebugTileStateDatabase`.
* [x] Write unit tests
* [x] `computeLoadProgress` needs to take view groups into account. It should be possible to provide an overload that reports the progress for a particular view group, and one that reports overall progress of all view groups.
* [x] Update tile selection algorithm docs based on these changes.
* [x] Credit handling
* [x] Check that `updateViewGroupOffline` still works, and add a view group version of it.
* [x] Consider whether `Tile::_doNotUnloadSubtreeCount` and `_referenceCount` can be combined into one. (#1156)
* [x] Consider whether `LoadedTileEnumerator` should continue to enumerate the root tile when no tiles at all are loaded.
* [x] Consider whether `updateView` should be deprecated, and users told to use `updateViewGroup` and `loadTiles` instead.
* [x] Check whether tile fade in/out is working. It probably needs to switch to a fade value per view instead of per tile.

kring commented:
In some quick profiling, it looks like tile selection state tracking is responsible for ~20% of total time spent in `updateView`. That's much higher than I'd like.

How much time was it taking before? It's hard to say exactly, because it's inlined. But previously this was little more than accessing a field on the `Tile`, which should already be pretty warm in the CPU cache. I expect it was very nearly zero.

j9liu commented:
@kring are there any testing instructions for this PR (e.g. in Unreal / Unity?)

kring commented:
@j9liu I'm close to having a not-totally-terrible implementation of multiple viewports in Unreal that you can use to test, but ran out of steam for tonight. It's probably reasonable to wait until tomorrow to review this so that I can get that in.

j9liu commented on Cesium3DTilesSelection/include/Cesium3DTilesSelection/TilesetViewGroup.h:
```diff
@@ -0,0 +1,230 @@
+ * {@link Tileset::updateViewGroup} to select tiles suitable for rendering the
+ * tileset from a given view or set of views.
+ */
+class CESIUM3DTILESSELECTION_API TilesetViewGroup final
```
I probably should have looked at this file earlier because it's the whole point of the PR... but I was surprised that this "view group" doesn't actually have any sort of "view" objects. So I'm not sure I understand the relationship between views in a runtime client -- i.e. literal cameras -- and a view group here. How would client views get associated with a specific view group, or multiple?

j9liu replied:
Maybe this is my <7hrs of sleep talking, but if I replace the name with just `TilesetView` it clicks more easily in my head...

j9liu replied:
Ok, after re-reading the README description for this class, I think my source of confusion was that I expected the `ViewState`s to be associated with one group, which is not necessarily true. The `ViewState`s (which are what "views" conceptually are, from my understanding), are decoupled from the `TilesetViewGroup`, because a client could dynamically change the `ViewState`s that are viewing a tileset...

The `TilesetViewGroup` is not a literal group of views, but rather manages the state of tiles in the context of some group of views... but it also doesn't have any actual internal knowledge of what views influenced it, it simply exists and the client has to keep track of it... is that a fair assessment? 🤣

kring replied:
Haha yes that's right. Just like `Tileset` previously didn't have a list of views.

There's no reason to force the user to manage a list of ViewStates owned by the view group and/or tileset when they're almost certainly going to create them from their own view definitions every frame anyway.

j9liu replied:
Gotcha, I think I get it! I might propose tweaking the doc for a bit more clarity, but it's mostly my brain being a little too literal and silly 😋

j9liu commented on Cesium3DTilesSelection/src/Tileset.cpp:
```diff
@@ -333,48 +328,63 @@ void Tileset::_updateLodTransitions(
 
 const ViewUpdateResult&
 Tileset::updateViewOffline(const std::vector<ViewState>& frustums) {
```
I never learned the nuances behind `updateViewOffline`, but does this imply there should be an `updateViewGroupOffline` too?

j9liu reviewed (commented):
Thanks for tackling this feature @kring! I'm not done with the full review, but I had some questions, and I figure if I asked them earlier then the rest of the review could be more streamlined.

kring commented on CesiumUtility/include/CesiumUtility/CreditSystem.h:
```diff
@@ -29,6 +30,23 @@ struct CESIUMUTILITY_API Credit {
   Credit(size_t id_) noexcept : id(id_) {}
```
This didn't change. `Credit` has always (or at least for a long time, before this PR) been an opaque handle to a credit created from HTML.

kring commented on Cesium3DTilesSelection/src/Tile.cpp:
```diff
@@ -1,11 +1,15 @@
+#include "TilesetContentManager.h"
```
No, because it's in the `src` directory. Using quotes is correct here.

j9liu replied:
Durr, I forgot to remove this when I saw `TilesetContentManager` in src later down the line. 🤪

kring commented on Cesium3DTilesSelection/include/Cesium3DTilesSelection/TileLoadTask.h:
```diff
@@ -0,0 +1,78 @@
+   *
+   * Tiles with a _lower_ value for this property load sooner!
+   */
+  double priority;
```
Yeah, it's a little confusing, but flipping it isn't really doable because the priority values are unbounded. Unimportant tiles (such as tiles way off in the distance) can have an arbitrarily-large priority value. In any case, this hasn't changed in this PR.

kring commented on Cesium3DTilesSelection/src/TilesetContentManager.h:
```diff
@@ -153,12 +150,22 @@ class TilesetContentManager
+  void markTileNowUnused(const Tile& tile);
+
+  void
+  unloadCachedBytes(int64_t maximumCachedBytes, double timeBudgetMilliseconds);
```
This isn't new, but good idea, I'll add some comments.

kring commented on Cesium3DTilesSelection/include/Cesium3DTilesSelection/Tile.h:
```diff
@@ -627,8 +626,18 @@ class CESIUM3DTILESSELECTION_API Tile final {
+  // external tileset (if any) from being unloaded from the tree. A non-zero
+  // count here prevents this Tile instance from being destroyed, but it does
+  // not prevent its content from being unloaded.
+  int32_t _doNotUnloadSubtreeCount;
```
Well, when greater than zero, it literally means we can't unload the subtree rooted at this tile. :) 
`_doNotUnloadSubtreeCount` isn't new in this PR, and `_viewGroupReferenceCount` is not correct because it also applies to height queries and potentially other types of queries in the future. But I'll see if I can add some clarity here.

kring replied:
I think I see a way to eliminate one of these reference counts completely. I'll have a try. Thanks for the push on it.

kring commented on Cesium3DTilesSelection/include/Cesium3DTilesSelection/LoadedTileEnumerator.h:
```diff
@@ -0,0 +1,198 @@
+ * @brief A "virtual collection" that allows enumeration through the loaded
+ * tiles in a subtree rooted at a given {@link Tile}.
+ *
+ * For the purposes of this enumeration, a loaded tile is one that is in a
+ * {@link TileLoadState} other than {@link TileLoadState::Unloaded}, or that
+ * has any children (or other descendants) that meet this criteria. We check the
+ * latter criteria by looking at {@link Tile::getDoNotUnloadSubtreeCount}.
```
Yep, as written, the root tile (if it exists) is _always_ enumerated. We could change this. I don't think it makes any practical difference at all, but I guess maybe that would be more sensible / less surprising?

kring commented on Cesium3DTilesSelection/include/Cesium3DTilesSelection/Tileset.h:
```diff
@@ -236,11 +251,35 @@ class CESIUM3DTILESSELECTION_API Tileset final {
+   * {@link getRootTileAvailableEvent}), the returned instance will remain
+   * valid until the tileset is destroyed.
+   *
+   * While the returned enumerator itself will remain valid as long as the
+   * `Tileset` does, a given iteration may be invalidated by any operation that
+   * modifies the {@link Tile} hierarchy.
```
Yeah that's what this comment is saying. Modifying the hierarchy during enumeration will result in undefined behavior. Just like erasing an element while iterating a vector would. But restarting a new enumeration after modifying the hierarchy is fine. There shouldn't be any cases in the code where we risk modification of the hierarchy during enumeration, but definitely call it out if you see one because that would be a bad bug.

kring commented on Cesium3DTilesSelection/include/Cesium3DTilesSelection/Tileset.h:
```diff
@@ -220,6 +230,11 @@ class CESIUM3DTILESSELECTION_API Tileset final {
   /**
    * @brief Updates this view, returning the set of tiles to render in this
    * view.
+   *
+   * Calling this method is equivalent to calling {@link updateViewGroup} with
+   * the default view group ({@link getDefaultViewGroup}) and then calling
+   * {@link loadTiles}.
+   *
```
Yep this was purely to avoid a breaking change, because most clients (including our own) aren't going to bother with view groups.

If the name is too confusing and we're going to deprecate it, I'd probably not replace it with anything. Clients will just be prompted by the deprecation message to reply `updateView` with `updateViewGroup` plus `loadTiles`.

kring commented:
Thanks for the review @j9liu! I responded to a few of the questions/comments. For the rest I'll make the requested changes today.

kring commented:
I opened CesiumGS/cesium-unreal#1652 to let us play with multiple viewports in Unreal.

kring commented on CesiumUtility/include/CesiumUtility/TreeTraversalState.h:
```diff
@@ -0,0 +1,406 @@
+namespace CesiumUtility {
+
+/**
+ * @brief Associates state (arbitrary data) with each node during partial,
+ * depth-first traversal of a tree. Then, during a later traversal of a
+ * potentially different subset of the same tree, the state previously
+ * associated with each node can be looked up.
```
Sorry, that doesn't seem like an improvement to me. It doesn't represent the result, really, at least not in the way a map of node -> state would. It's a more active partner in the traversal rather than a result. And I think it's less clear that only the previous state for the "current" node can be looked up efficiently.

But I'm sure there's room for improvement here. Was the main issue that the original was difficult to understand?

j9liu replied:
Don't apologize -- I see your point. I think this was one of the first files I looked at, so a lot of the code hadn't clicked for me at that point.

>  And I think it's less clear that only the previous state for the "current" node can be looked up efficiently.

I think that's slightly less clear from the wording, so one improvement could be to point that out. But otherwise I should wait until the real end of the review to suggest anything, so I actually know what I am talking about 😁

kring commented:
Thanks again for the review @j9liu. I've addressed most of your comments, and turned the remaining ones into to-do items in the top post.

j9liu commented on CesiumUtility/include/CesiumUtility/CreditReferencer.h:
```diff
@@ -0,0 +1,110 @@
+/**
+ * @brief Provides a way to reference a set of credits in a
+ * {@link CreditSystem} so that the references can easily be released later.
+ */
```
Maybe add a clause like "Multiple CreditReferencers may be used on the same credit system to track separate sets of references -- e.g., two sets of credits from different frames."

That could be deduced from the way `CreditReferencer` is used, but it still helps to state the rationale in the class's own documentation.

j9liu commented on CHANGES.md:
```diff
@@ -1,5 +1,39 @@
+
+##### Breaking Changes :mega:
+
+- Removed `addCreditToFrame`, `startNextFrame`, `getCreditsToShowThisFrame`, and `getCreditsToNoLongerShowThisFrame` from `CreditSystem`. `CreditSystem` no longer has a notion of a "frame". Instead, credits are included and excluded by calling `addCreditReference` and `removeCreditReference`. A snaphot of the current state can be obtained by calling `getSnapshot`, and it includes both the current ...
```
`TilesetExternals` has a comment for its `pCreditSystem` that references "each frame", so maybe that wording should changed too?

j9liu commented on CesiumUtility/test/TestCreditReferencer.cpp:
```diff
@@ -0,0 +1,122 @@
+      const CreditsSnapshot& snapshot3 = pCreditSystem->getSnapshot();
+      REQUIRE(snapshot3.currentCredits.size() == 2);
+
+      const CreditsSnapshot& snapshot4 = pCreditSystem->getSnapshot();
+      REQUIRE(snapshot4.currentCredits.size() == 2);
```
Remove? (unless we're purposely testing snapshots that are consecutive with no changes)

j9liu commented on CesiumUtility/include/CesiumUtility/TreeTraversalState.h:
```diff
@@ -0,0 +1,406 @@
+   * etc.) of the current node that has been traversed so far in the current
+   * traversal.
+   *
+   * If the current node's children were not traversed in the previous
```
?

```suggestion
   * If the current node's children were not traversed in the current
```

j9liu commented on CesiumUtility/test/TestTreeTraversalState.cpp:
```diff
@@ -0,0 +1,481 @@
+     *     e   f
+     */
+    // clang-format off
+    traversalState.beginNode(&a);
```
Definitely a nitpick but perhaps nesting scopes with `{}` could help with the readability / formatting?

kring replied:
This is a good idea, but I tried it out on one of the blocks, and I think I prefer it as it is. The braces are nice because clang-format knows how to handle format it, but they add a like of vertical space and I don't think make it particularly easier to read. Or maybe I'm just lazy and don't feel like doing it everywhere. 😁

j9liu reviewed (commented):
Finally finished the second pass @kring ! Mostly doc / question stuff this time as well. Thanks for addressing the comments from before :smile:

kring commented on Cesium3DTilesSelection/src/TilesetViewGroup.cpp:
```diff
@@ -0,0 +1,257 @@
+    this->_mainThreadLoadQueue.emplace_back(task);
+  } else if (
+      pTile->getState() == TileLoadState::ContentLoading ||
+      pTile->getState() == TileLoadState::Unloading) {
+    ++this->_tilesAlreadyLoading;
```
This `_tilesAlreadyLoading` is used for progress tracking. And when we hit this branch, we have a tile that needs to be loaded. To illustrate: if this were the only tile this view needed, the progress wouldn't be 100% yet, right? Furthermore, it's a tile we can't add to either other queue because it's already transitioning.

So the behavior is correct here. I guess I'm happy to squint a little bit and say that "already loading" can include "finishing unloading so it can be reloaded."

But maybe `_tilesAlreadyLoadingOrUnloading` would be clearer? I'll rename it.

kring commented on Cesium3DTilesSelection/src/Tileset.cpp:
```diff
@@ -1152,34 +1117,23 @@ bool Tileset::_loadAndRenderAdditiveRefinedTile(
 // TODO This function is obviously too complex. The way how the indices are
 // used, in order to deal with the queue elements, should be reviewed...
```
Well... I'd be fair to say it's still too complex, probably. 😆 But I don't think the comment is adding anything, so I'll remove it.

kring commented:
Thanks for the second round of reviews @j9liu! I think I've addressed all your feedback. I still need to address the few things listed in the to-do list in the top post, though.

kring commented:
I believe that this PR, plus #1156 and #1160, address all the review feedback and this is ready to go.

j9liu reviewed (approved):
Thanks @kring for bearing with all my feedback :smile: There might be one outstanding comment in `TilesetHeightQuery.cpp`, but it's not a big deal, so don't sweat to address it.

I'm approving this PR, though I'm holding off the merge since it's a big change, and we only have a week before the next release. If we feel more confident to include it, let me know!

j9liu commented:
Thanks @kring ! Merging once CI passes
